design-proposals/klip-3-with-output-schema.md

```diff
@@ -0,0 +1,90 @@
+
+## Motivation and background
+
+Currently connect schemas pushed to the schema registry are inferred by KSQL which means all field names become uppercase (except you explicitly write SELECT id as `id`, ...), all fields become optional and Maps become arrays of MapEntries which is a result of all fields becoming optional. This behavior is intended and not a bug.
+
+But sometimes it's undesirable. Let's say you want to filter a few properties of an existing topic containing avro data and write the result into a new topic, you expect that the new topic has the same field definitions than the filtered one. But that's not the case. Even worse: In my specific case I was not able to use KSQL and had to go for something else, because the new map schema was not compatible to the one I expected which caused issues in the application using the target topic. So we should give more control over the resulting data schema.
+
+Example:
+
+```
+CREATE TABLE FOO
+WITH (KAFKA_TOPIC='foo', VALUE_FORMAT='AVRO', OUTPUT_SCHEMA='bar')
+AS SELECT * FROM BAR;
+```
+
+## What is in scope
+
+I suggest to introduce a new ``WITH`` property called ```OUTPUT_SCHEMA``` that can be used to explain the query which exact schema to use for the resulting topic and its data. So this new property is only available in ```CREATE TABLE/STREAM AS SELECT``` clauses.
+
+## What is not in scope
+
+We will not add support for any other ```VALUE_FORMAT``` than ```AVRO```, because the schema evolution issue only affects AVRO.
+
+## Value/Return
+
+The user can decide which resulting schema to go for. This means that many will no longer have a headache when it comes to deriving schemas. Sometimes users don't want optional fields because it's breaking their type validation if the schema registry is put in a larger context than just ksql - for example a REST API. Mapping maps to arrays of MapEntries is not a valid type cast even if there are good reasons to do so, for example for joined streams. So some use cases are currently not implementable, for example cloning a topic using KSQL. I personally cannot use the current version of KSQL for my use case.
+
+So the overall value is that you can use KSQL better in combination with Kafka Rest Proxy.
+
+## Public APIS
+
+The API is extended by a new ```WITH``` property called ```OUTPUT_SCHEMA``` that is only available for ```CREATE STREAM/TABLE AS SELECT``` and only in combination with the ```VALUE_FORMAT``` ```AVRO```.
+
+## Design
+
+The new ```WITH``` property is analyzed by the Analyzer and stored in the Query Analysis. That information is processed by the query engine and used by the AvroSerializer (KsqlAvroTopicSerDe). KsqlAvroTopicSerDe uses the schema registry client to get the latest metadata of the related output schema and uses this schema instead of the current one.
```

Comment:
Won't think require KSQL to validate that the selected schema is compatible with the actual data being produced? Is there a schema registry API to do this?

Reply:
Yes it must be validated if the selected schema is be compatible to the actual data being produced. We can check the original schema against the selected one by using this method:

https://github.com/confluentinc/schema-registry/blob/master/client/src/main/java/io/confluent/kafka/schemaregistry/client/SchemaRegistryClient.java#L55

Subject is the <schema>-value, schema the selected one. If it's returning false we have to throw an exception telling the user that the new schema does not fit to the generated data.

Reply:
@apurvam How shall we continue? Should I adjust the text?